### PR TITLE
Flatten layer options and improve sourceOptions naming

### PR DESCRIPTION
**Pull request by alex3165**
Should fix https://github.com/alex3165/react-mapbox-gl/issues/370 and https://github.com/alex3165/react-mapbox-gl/issues/310

TODO:
- [x] Update props interface / behavior
- [x] Write unit test accordingly
- [x] Update documentation

### CODE REVIEW

File: docs/API.md [hunk 1]
```diff
@@ -143,15 +143,23 @@ import { Layer } from "react-mapbox-gl";
...
-- **paint**: `Object` Mapbox paint object passed down to mapbox `addLayer` method [mapbox paint api](https://www.mapbox.com/mapbox-gl-style-spec/#layer-paint)
-- **sourceOptions**: `Object` Options object merged to the object used when calling `GeoJSONSource` method
-- **layerOptions**: `Object` Passed down to the layer object when setting it out.
+  - `fill-extrusion`, Include a Mapbox fill extrusion layer
+  - `background`, Include a Mapbox background layer
+- **layout**: `Object` Mapbox layout object passed down to mapbox `map.addLayer` method [mapbox layout api](https://www.mapbox.com/mapbox-gl-style-spec/#layer-layout)
+- **paint**: `Object` Mapbox paint object passed down to mapbox `map.addLayer` method [mapbox paint api](https://www.mapbox.com/mapbox-gl-style-spec/#layer-paint)
+- **geoJSONSourceOptions**: `Object` Source options merged to object passed to `map.addSource`
+- **metadata**: `Object` metadata parameter passed to `map.addLayer`
+- **layerRef**: `String` ref parameter passed to `map.addLayer`
```
[the remark this answers is not in the document]
Comment (alex3165) [reply docs/API.md]: You can't pass ref as props of a component

File: src/layer.ts [hunk 2]
```diff
@@ -37,18 +37,26 @@ export type ImageDefinitionWithOptions = [
 ];
 
 export interface LayerCommonProps {
-  type?: 'symbol' | 'line' | 'fill' | 'circle' | 'raster';
+  type?: 'symbol' | 'line' | 'fill' | 'circle' | 'raster' | 'fill-extrusion' | 'background';
```
[the remark this answers is not in the document]
Comment (alex3165) [reply src/layer.ts]: Alright I will do the upgrade in the same PR

[the remark this answers is not in the document]
Comment (alex3165) [reply src/layer.ts]: We also need a heatmap example in the demo page I think

File: docs/API.md [hunk 3]
```diff
@@ -143,15 +143,23 @@ import { Layer } from "react-mapbox-gl";
...
+- **sourceLayer**: `String` source-layer parameter passed to `map.addLayer`
+- **minZoom**: `Number` minzoom parameter passed to `map.addLayer`
+- **maxZoom**: `Number` maxzoom parameter passed to `map.addLayer`
+- **interactive**: `Boolean` interactive parameter passed to `map.addLayer`
```
[the remark this answers is not in the document]
Comment (alex3165) [reply docs/API.md]: 👍

File: docs/API.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (Wykks) [reply docs/API.md]: I mean, I think `ref` doesn't exist anymore :D
https://www.mapbox.com/mapbox-gl-js/style-spec/#layers

[the remark this answers is not in the document]
Comment (alex3165) [reply docs/API.md]: oh right!

File: src/layer.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (alex3165) [reply src/layer.ts]: will add heatmap with upgrade of mapbox-gl

**Pull request merged**